File: Ahorn/entities/smwCheckpoint.jl
module FlushelineSMWCheckpoint

using ..Ahorn, Maple

@mapdef Entity "vitellary/smwcheckpoint" Checkpoint(x::Integer, y::Integer, height::Integer=16, fullHeight::Bool=false)

const placements = Ahorn.PlacementDict(
    "SMW Checkpoint (Crystalline)" => Ahorn.EntityPlacement(
        Checkpoint,
        "rectangle",
        Dict{String, Any}(),
        function(entity)
            entity.data["nodes"] = [(Int(entity.data["x"]), Int(entity.data["y"]) + 8)]
        end
    )
)

Ahorn.nodeLimits(entity::Checkpoint) = 1, 1

function Ahorn.selection(entity::Checkpoint)
    x, y = Ahorn.position(entity)
    height = Int(get(entity.data, "height", 16))
    node = get(entity.data, "nodes", ())[1]
    
    return [Ahorn.Rectangle(x, y, 16, height), Ahorn.Rectangle(x+2, node[2], 12, 4)]
end

Ahorn.minimumSize(entity::Checkpoint) = 0, 16
Ahorn.resizable(entity::Checkpoint) = false, true

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::Checkpoint, room::Maple.Room)
    x, y = Ahorn.position(entity)
    height = Int(get(entity.data, "height", 16))
    node = get(entity.data, "nodes", ())[1]
    barHeight = node[2]
    
    for i in 0:(floor(Int, height / 8) - 1)
        if i == 0
            Ahorn.drawImage(ctx, "objects/smwCheckpoint/bars", x, y, 0, 0, 4, 8)
        else
            Ahorn.drawImage(ctx, "objects/smwCheckpoint/bars", x, y + i*8, 0, 8, 4, 8)
        end
    end
    Ahorn.drawImage(ctx, "objects/smwCheckpoint/cp", x+2, barHeight)
    for i in 0:(floor(Int, height / 8) - 1)
        if i == 0
            Ahorn.drawImage(ctx, "objects/smwCheckpoint/bars", x + 12, y, 4, 0, 4, 8)
        else
            Ahorn.drawImage(ctx, "objects/smwCheckpoint/bars", x + 12, y + i*8, 4, 8, 4, 8)
        end
    end
end

end

File: Loenn/entities/smw_checkpoint.lua
local utils = require "utils"
local drawableSprite = require "structs.drawable_sprite"

local smwCheckpoint = {}

smwCheckpoint.name = "vitellary/smwcheckpoint"
smwCheckpoint.placements = {
    {
        name = "checkpoint",
        data = {
            height = 16,
            fullHeight = false,
        },
    },
}
smwCheckpoint.nodeLimits = {1, 1}
smwCheckpoint.nodeVisibility = "never"
smwCheckpoint.canResize = {false, true}
smwCheckpoint.minimumSize = {0, 16}

function smwCheckpoint.sprite(room, entity)
    local sprites = {}
    
    for i=0, math.floor(entity.height / 8) - 1 do
        local quad_x, quad_y = 0, (i == 0) and 0 or 8
        local sprite = drawableSprite.fromTexture("objects/smwCheckpoint/bars", entity)
        sprite.y += i*8
        sprite:useRelativeQuad(quad_x, quad_y, 4, 8)
        table.insert(sprites, sprite)
    end

    local cp = drawableSprite.fromTexture("objects/smwCheckpoint/cp", entity)
    cp:setPosition(entity.x + 2, entity.nodes[1].y)
    cp:setJustification(0, 0.5)
    table.insert(sprites, cp)

    for i=0, math.floor(entity.height / 8) - 1 do
        local quad_x, quad_y = 4, (i == 0) and 0 or 8
        local sprite = drawableSprite.fromTexture("objects/smwCheckpoint/bars", entity)
        sprite:addPosition(12, i*8)
        sprite:useRelativeQuad(quad_x, quad_y, 4, 8)
        table.insert(sprites, sprite)
    end

    return sprites
end

function smwCheckpoint.selection(room, entity)
    local x, y, w, h = entity.x or 0, entity.y or 0, 16, entity.height or 16
    return utils.rectangle(x, y, w, h), {utils.rectangle(x+2, (entity.nodes[1].y or y) - 2, 12, 4)}
end

return smwCheckpoint